java-squid/src/main/java/org/sonar/java/se/ConstraintManager.java

```diff
@@ -0,0 +1,253 @@
+          return SymbolicValue.TRUE_LITERAL;
+        } else if (val.equals(SymbolicValue.TRUE_LITERAL)) {
+          return SymbolicValue.FALSE_LITERAL;
+        }
+        // if not tied to a concrete value, create symbolic value with no constraint for now.
+        // TODO : create constraint between expression and created symbolic value
+      default:
+    }
+    return createSymbolicValue(syntaxNode);
+  }
+
+  /**
+   * Remove parenthesis and type cast.
+   */
+  private static Tree skipTrivial(Tree givenExpression) {
+    Tree expression = givenExpression;
+    do {
+      switch (expression.kind()) {
+        case PARENTHESIZED_EXPRESSION:
+          expression = ((ParenthesizedTree) expression).expression();
+          break;
+        case TYPE_CAST:
+          expression = ((TypeCastTree) expression).expression();
+          break;
+        default:
+          return expression;
+      }
+    } while (true);
+  }
+
+  public boolean isNull(ProgramState ps, SymbolicValue val) {
+    return NullConstraint.NULL.equals(ps.constraints.get(val));
+  }
+
+  public boolean isConstrained(ProgramState ps, SymbolicValue val) {
+    return ps.constraints.containsKey(val);
+  }
+
+  public Pair<ProgramState, ProgramState> assumeDual(ProgramState programState, Tree givenCondition) {
+    // FIXME condition value should be evaluated to determine if it is worth exploring this branch. This should probably be done in a
+    // dedicated checker.
+    Tree condition = skipTrivial(givenCondition);
+    switch (condition.kind()) {
+      case INSTANCE_OF:
+        InstanceOfTree instanceOfTree = (InstanceOfTree) condition;
+        SymbolicValue exprValue = eval(programState, instanceOfTree.expression());
+        if (isNull(programState, exprValue)) {
+          return new Pair<>(programState, null);
+        }
+        // if instanceof is true then we know for sure that expression is not null.
+        return new Pair<>(programState, setConstraint(programState, exprValue, NullConstraint.NOT_NULL));
+      case EQUAL_TO:
+        BinaryExpressionTree equalTo = (BinaryExpressionTree) condition;
+        SymbolicValue equalToLhs = eval(programState, equalTo.leftOperand());
+        SymbolicValue equalToRhs = eval(programState, equalTo.rightOperand());
+        if (isNull(programState, equalToLhs)) {
+          ProgramState stateNull = setConstraint(programState, equalToRhs, NullConstraint.NULL);
+          ProgramState stateNotNull = setConstraint(programState, equalToRhs, NullConstraint.NOT_NULL);
+          return new Pair<>(stateNotNull, stateNull);
+        } else if (isNull(programState, equalToRhs)) {
+          ProgramState stateNull = setConstraint(programState, equalToLhs, NullConstraint.NULL);
+          ProgramState stateNotNull = setConstraint(programState, equalToLhs, NullConstraint.NOT_NULL);
+          return new Pair<>(stateNotNull, stateNull);
+        }
+        break;
+      case NOT_EQUAL_TO:
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce this switch case number of lines from 13 to at most 5, for example by extracting code into methods. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1151)